subsys/caf/modules/sensor_data_aggregator.c

```diff
@@ -0,0 +1,209 @@
+{
+	for (size_t i = 0; i < agg->buf_count; i++) {
+		if (agg->agg_buffers[i].busy == false) {
+			return &agg->agg_buffers[i];
+		}
+	}
+	return NULL;
+}
+
+static struct aggregator *find_aggregator(const char *descr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(aggregators); i++) {
+		if (descr == aggregators[i].descr) {
+			return &aggregators[i];
+		}
+	}
+	return NULL;
+}
+
+static int clean_agg_buffer(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	__ASSERT_NO_MSG(ab);
+
+	LOG_INF("Cleaning, sample_cnt %d", ab->sample_cnt);
+	ab->sample_cnt = 0;
+	ab->busy = false;
+	if (agg->active_buf == NULL) {
+		agg->active_buf = ab;
+	}
+	return 0;
+}
+
+static void send_agg_buff_data(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	ab->busy = true;
+	agg->active_buf = find_non_busy_buffer(agg);
+
+	struct sensor_aggregator_event *event = new_sensor_aggregator_event();
+
+	event->buf = ab->data;
+	event->sample_cnt = ab->sample_cnt;
+	event->sensor_status = agg->sensor_state;
+	event->descr = agg->descr;
+	EVENT_SUBMIT(event);
+}
+
+static int add_sample_to_aggregator(struct sensor_event *event, struct aggregator *agg)
+{
+	if (event->dyndata.size != agg->sensor_data_size) {
+		return -EBADMSG;
+	}
+	if (!agg->active_buf) {
+		return -ENOMEM;
+	}
+	struct aggregator_buffer *ab = agg->active_buf;
+
+	size_t pos = ab->sample_cnt * agg->sensor_data_size;
+	size_t avail = agg->buf_len - pos;
+
+	if (avail < agg->sensor_data_size) {
+		__ASSERT_NO_MSG(false);
+		return -ENOMEM;
+	}
+
+	memcpy(&ab->data[pos], event->dyndata.data, event->dyndata.size);
+	ab->sample_cnt++;
+	avail -= event->dyndata.size;
+
+	if (avail < agg->sensor_data_size) {
+		send_agg_buff_data(agg, ab);
+	}
+
+	return 0;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_sensor_event(eh)) {
+		struct sensor_event *event = cast_sensor_event(eh);
+		struct aggregator *agg = find_aggregator(event->descr);
+
+		if (agg) {
+			int err = add_sample_to_aggregator(event, agg);
+
+			if (err) {
+				LOG_ERR("Error code: %d", err);
+			}
+		} else {
+			LOG_WRN("Dropped event: %s. Found no adequate aggregator.", event->descr);
+		}
+		return false;
+	}
+
+	if (is_sensor_aggregator_release_buffer_event(eh)) {
+		const struct sensor_aggregator_release_buffer_event *event =
+				cast_sensor_aggregator_release_buffer_event(eh);
+		struct aggregator *agg = find_aggregator(event->descr);
+
+		if (agg) {
+			for (size_t i = 0; i < agg->buf_count; i++) {
+				if ((agg->agg_buffers[i].data) == event->buf) {
+					clean_agg_buffer(agg, &agg->agg_buffers[i]);
+					break;
+				}
+			}
+		}
+
+		return false;
+	}
+
+	if (is_sensor_state_event(eh)) {
+		struct sensor_state_event *event = cast_sensor_state_event(eh);
+		struct aggregator *agg = find_aggregator(event->descr);
+
+		if (agg) {
+			agg->sensor_state = event->state;
+			if (agg->active_buf->sample_cnt != 0) {
+				send_agg_buff_data(agg, agg->active_buf);
+			}
```

Comment:
If a sensor changes its state and no data were stored in aggregator, this event would be silently dropped. Imagine situation:

1. Sensor has trigger enabled and can put core to sleep if nothing happens after lets say 8 samples.
2. Sensor is `ACTIVE` and sampled 8/30 times -> 8 samples in aggregator buffer.
3. Trigger is enabled, sensor transition to `SLEEP` - >aggregator sends out 8 samples and `state=SLEEP`.
4. some time elapses...
5. Trigger wakes up the platform. Sensor samples again (`state=ACTIVE`). Samples are constantly put to aggregator.
6. After lets say 20 samples trigger is activated once again. You hit this function.
7. `sensor_state=SLEEP` 20/30 samples from aggregator send out.
8. Listener to `sensor_aggregator_event` received 20 additional samples when sensor was in `state=SLEEP`.

That's why i'm not sure if combining those two is good. I might be picky just wanted to share this and see your opinion.
That could possibly cause strange behavior if the listener to `sensor_aggregator_event` does not listen to `sensor_state` as it assumes this is the only way of knowing sensor state. Transition between ACTIVE-SLEEP-ACTIVE-SLEEP is hidden.

Reply:
Yes, that's right, good point.
Maybe as a solution we should remove this check and send sensor_aggregator_event  no matter if there is data in aggregator:
			if (agg->active_buf->sample_cnt != 0) {